libs/utils/include/utils/RangeMap.h

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright (C) 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TNT_UTILS_RANGEMAP_H
+#define TNT_UTILS_RANGEMAP_H
+
+#include <utils/Panic.h>
+#include <utils/Range.h>
+#include <utils/debug.h>
+
+#include <map>
+
+namespace utils {
+
+/**
+ * Sparse container for a series of ordered non-overlapping intervals.
+ *
+ * RangeMap has a low memory footprint if it contains fairly homogeneous data. Internally, the
+ * intervals are automatically split and merged as elements are added or removed.
+ *
+ * Each interval maps to an instance of ValueType, which should support cheap equality checks
+ * and copy assignment. (simple concrete types are ideal)
+ *
+ * KeyType should support operator< because intervals are internally sorted using std::map.
+ */
+template<typename KeyType, typename ValueType>
+class RangeMap {
```

Comment:
note: when actually using this class, it *might* be good to explicitly instantiate it, because it looks like all these methods could generate a lot of code.

One question I have is, is it possible to refactor this so either:
- some code could be moved into a .cpp (i.e. code that doesn't depend on the templated types)
- or, make the templated class just a wrapper (for type safety) around a generic class

Reply:
Hmm I tried but wasn't able to move any code into a cpp file, since all methods use template types.  Keeping it all in the header might not be so bad, at least it ensures that we don't bloat libutils for clients who don't need RangeMap.

FWIW, the Vulkan backend will likely instantiate this with KeyType=uint32 and ValueType=VkImageLayout.